src/lib/cache/gdfs_cache.hpp

```diff
@@ -145,15 +172,27 @@ class GDFSCache : public AbstractCacheImpl<Key, Value> {
     return entry.frequency;
   }
 
+  CacheSnapshot snapshot() const {
+    std::shared_lock<std::shared_mutex> lock(_mutex);
+    CacheSnapshot _map_copy;
+    for (auto it = _map.begin(); it != _map.end(); it++) {
+      const auto [key, entry] = *it;
+      _map_copy[key] = *entry;
+    }
+    return _map_copy;
+  }
+
  protected:
   // Priority queue to hold all elements. Implemented as max-heap.
   boost::heap::fibonacci_heap<GDFSCacheEntry> _queue;
 
   // Map to point towards element in the list.
   CacheMap _map;
 
+  mutable std::shared_mutex _mutex;
+
   // Inflation value that will be updated whenever an item is evicted.
-  double _inflation;
+  std::atomic<double> _inflation;
```

Comment:
We should document how we achieve thread-safety. People might be surprised that we mix locking and atomics. Also, a short comment about why we decided for locked unordered_maps instead of lock-free data structures would be helpful.

Reply:
Basically, this could also be a normal double and we lock the inflation() function. If we lock all the other functions as well, this should be fine.

Reply:
It looks like most of these getters are just used for tests, anyway. Should we simply friend the test class and get rid of the accessors altogether?

Reply:
I guess this would make the interface of the cache much cleaner. What about priority and frequency? Should they remain at all or public?

Reply:
A question for the people that look at the cache, again

Reply:
I don't mind removing some getters. For me as a cache user, I am only interested in the frequency in the snapshot. Nowhere else.
And I don't know what the priority really does, so I can't say anything about that.